Tests/AsyncQueueTests/AsyncQueueTests.swift

```diff
@@ -26,8 +26,149 @@ import XCTest
+    }
+
+    // MARK: Behavior Tests
+
+    func test_async_sendsEventsInOrder() {
+        let counter = Counter()
+        for iteration in 1...1_000 {
+            systemUnderTest.async {
+                await counter.incrementAndExpectCount(equals: iteration)
+            }
+        }
+    }
+
+    func test_async_executesAsyncBlocksSeriallyAndAtomically() {
+        let unsafeCounter = UnsafeCounter()
+        for iteration in 1...1_000 {
+            systemUnderTest.async {
+                unsafeCounter.incrementAndExpectCount(equals: iteration)
+            }
+        }
+    }
+
+    func test_async_isNotReentrant() async {
+        let counter = Counter()
+        await systemUnderTest.await { [systemUnderTest] in
+            systemUnderTest.async {
+                await counter.incrementAndExpectCount(equals: 2)
+            }
+            await counter.incrementAndExpectCount(equals: 1)
+            systemUnderTest.async {
+                await counter.incrementAndExpectCount(equals: 3)
+            }
+        }
+    }
+
+    func test_async_retainsReceiverUntilFlushed() async {
+        var systemUnderTest: AsyncQueue? = AsyncQueue()
+        let counter = Counter()
+        let expectation = self.expectation(description: #function)
+        await withThrowingTaskGroup(of: Void.self) { taskGroup in
+            let foreverSleep = Task {
+                try await Task.sleep(nanoseconds: UInt64.max)
+            }
+            taskGroup.addTask {
+                try await foreverSleep.value
+            }
+            systemUnderTest?.async {
+                // Make the queue wait.
+                try? await foreverSleep.value
+                await counter.incrementAndExpectCount(equals: 1)
+            }
+            systemUnderTest?.async {
+                // This async task should not execute until the sleep is cancelled.
+                await counter.incrementAndExpectCount(equals: 2)
+                expectation.fulfill()
+            }
+            // Nil out our reference to the queue to show that the enqueued tasks will still complete
+            systemUnderTest = nil
+            // Cancel the sleep timer to unlock the remaining enqueued tasks.
+            foreverSleep.cancel()
+
+            await waitForExpectations(timeout: 1.0)
+        }
+    }
+
+    func test_await_sendsEventsInOrder() async {
+        let counter = Counter()
+        for iteration in 1...1_000 {
+            systemUnderTest.async {
+                await counter.incrementAndExpectCount(equals: iteration)
+            }
+
+            guard iteration % 25 == 0 else {
+                // Keep sending async events to the queue.
+                continue
+            }
+
+            await systemUnderTest.await {
+                let count = await counter.count
+                XCTAssertEqual(count, iteration)
+            }
+        }
+    }
+
+    func test_await_canReturn() async {
```

Comment:
This test and `test_await_canThrow` below are a little silly but I found them helpful for ensuring that the capability exists as I iterated on the queue's API.